### ISSUE TITLE
Consider moving emmet completion provider for css to css language service

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode#29113]
In the new emmet, you get css abbreviations inside the <style> tag. But you get it **everywhere** inside the style tag including the selectors which is not valid.

Contents of <style> tag should be re-parsed as css to find valid locations for emmet abbr expansion

Or the emmet completion provider for css should be moved to the css language service

**Comment by ramya-rao-a** [comment]
Moving emmet completion to css language service will also help in sorting the emmet suggestion among css suggestions See the discussion in #28933

**Comment by ramya-rao-a** [comment]
The sorting issue has been fixed with https://github.com/Microsoft/vscode-emmet-helper/commit/04593cad97afa5fa675dd868ffd9c8544b3a1612

Abbreviations like `ttu` (emmet gives`text-transform: uppercase`) or `tac` (emmet gives `text-align: center;` will now be shown on the top

**Comment by ramya-rao-a** [comment]
Things to do to get emmet support in css language service:

- Pass emmet settings from css client to css language service initially and on every change in these settings
- Currently the custom snippets is processed and cached. The cache is maintained in the `vscode-emmet-helper` module. We need to figure out the right place for this cache. Css client or language service or the helper module?
- Get UMD version of the emmet helper and hook it up with the language service

**Comment by ramya-rao-a** [comment]
Update: Most of the required work is done. See #41652
Waiting for next version of LSP that will have the change for new completion trigger kind, and so this is moved to the next iteration

**Comment by jens1o** [comment]
In case you need somebody for extensive testing, I'm available for you! :)

**Comment by ramya-rao-a** [comment]
Thanks @jens1o!
I'll mention you here in this issue when its ready which should be sometime in the second week of February.

**Comment by ramya-rao-a** [comment]
Calling for emmet champions, especially the ones who helped me a ton last year
@smlombardi, @jens1o, @vvs, @Niichie, @tklepzig

This feature is now complete and can be tested on latest Insiders dated 2/13
If you are using Win 64 bits, then please wait another day and test on the Insiders dated 2/14

Try out css emmet completions in css/less/scss files and also inside the `style` block in html files

**Comment by jens1o** [comment]
Yeeaaaahh. But now I have to wait until tomorrow. :/

**Comment by smlombardi** [comment]
`!important` does not offer a completion

**Comment by ramya-rao-a** [comment]
@smlombardi Fixed. Try again with tomorrow's build

**Comment by jens1o** [comment]
Working fine so far!

**Comment by ramya-rao-a** [comment]
This work has been reverted due to #44840 and #44352.
Re-opening and moving to March

**Comment by jens1o** [comment]
Will this affect the insiders release, too?

**Comment by ramya-rao-a** [comment]
@jens1o Yes, currently its reverted in Insiders release and will not be present in the upcoming 1.21 stable release.

**Comment by jens1o** [comment]
Alright, I'll keep my eye sharp to test it once it's back ;)

**Comment by ramya-rao-a** [comment]
https://github.com/Microsoft/vscode-css-languageservice/issues/81 and https://github.com/Microsoft/vscode-css-languageservice/issues/69 are 2 issues in the css language service that stops us from moving emmet to the css extension

**Comment by ramya-rao-a** [comment]
Below were the advantages we were hoping for by moving the emmet css completions to the css extension
- Emmet css completions in style attribute in html files
- Reusing the parsing done by the css extension and not relying on the css parser from emmet which tends to get much slower for larger files.

The second problem got solved when we implemented the partial parsing of css files last milestone. See https://github.com/Microsoft/vscode/issues/43470#issuecomment-376370823

The first problem can be solved in other ways. 

Since we have upstream issues that block this move and there isn't much to gain out of this move, we are closing this item and will continue to keep css emmet completions in the emmet extension itself
